logdetective/server/server.py

```diff
@@ -260,3 +273,167 @@ async def analyze_log_stream(build_log: BuildLog):
+    await asyncio.to_thread(job.artifacts, streamed=True, action=tempfile.write)
+    tempfile.seek(0)
+
+    failed_arches = {}
+    artifacts_zip = zipfile.ZipFile(tempfile, mode="r")
+    for zipinfo in artifacts_zip.infolist():
+        if zipinfo.filename.endswith("task_failed.log"):
+            # The koji logs store this file in two places: 1) in the
+            # directory with the failed architecture and 2) in the parent
+            # directory. We actually want to ignore the one in the parent
+            # directory, since the rest of the information is in the
+            # specific task directory.
+            # The paths look like `kojilogs/noarch-XXXXXX/task_failed.log`
+            # or `kojilogs/noarch-XXXXXX/x86_64-XXXXXX/task_failed.log`
+            path = PurePath(zipinfo.filename)
+            if len(path.parts) <= 3:
+                continue
+
+            # Extract the architecture from the immediate parent path
+            architecture = path.parent.parts[-1].split("-")[0]
+
+            # Open this file and read which log failed.
+            # The string in this log has the format
+            # `see <log> for more information`.
+            # Note: it may sometimes say
+            # `see build.log or root.log for more information`, but in
+            # that situation, we only want to handle build.log (for now),
+            # which means accepting only the first match for the regular
+            # expression.
+            with artifacts_zip.open(zipinfo.filename) as task_failed_log:
+                contents = task_failed_log.read().decode("utf-8")
+                match = FAILURE_LOG_REGEX.search(contents)
+                if not match:
+                    LOG.error(
+                        "task_failed.log does not indicate which log contains the failure."
+                    )
+                    raise SyntaxError(
+                        "task_failed.log does not indicate which log contains the failure."
+                    )
+                failure_log_name = match.group(1)
+
+            failed_arches[architecture] = PurePath(path.parent, failure_log_name)
+
+    if not failed_arches:
+        # No failed task found?
+        raise FileNotFoundError("Could not detect failed architecture.")
+
+    # First check if we only found one failed architecture
+    if len(failed_arches) == 1:
+        failed_arch = list(failed_arches.keys())[0]
+
+    else:
+        # We only want to handle one arch, so we'll check them in order of
+        # "most to least likely for the maintainer to have access to hardware"
+        # This means: x86_64 > aarch64 > ppc64le > s390x
+        if "x86_64" in failed_arches:
+            failed_arch = "x86_64"
+        elif "aarch64" in failed_arches:
+            failed_arch = "aarch64"
+        elif "ppc64le" in failed_arches:
+            failed_arch = "ppc64le"
+        elif "s390x" in failed_arches:
+            failed_arch = "s390x"
+        else:
```

Comment:
I wonder how will this code react to SRPM build failures

Reply:
I didn't find an example to test, so I probably need to create one...

I'm pretty sure that it still returns an archful log though. Even the noarch builds for python packages, et. al. do that for binary RPMs.

Reply:
Interesting, every buildsystem behaves differently.

If it doesn't work, we'll find out soon enough :)